### PR TITLE
feat: Gnark groth16 single block and composite block circuits

### PR DESCRIPTION
**Pull request by ziemen4**
- Create circuit for single block case
- Create circuit for composite block case
- Create common module for sha2 and utilities
- Add test inputs to input folder

Out of scope:
- Set up testing module to prevent setup and add tests for different use cases
- Refactor composite case to re-use single case circuit
- Refactor some longer constraint checks into methods

### CODE REVIEW

**Comment by alexhooketh** [comment]
👏👏no idea if that's good tho, i don't know go 
let me run this

File: gnark/common/utils.go
Comment (alexhooketh) [inline gnark/common/utils.go:1]: interesting stuff on the limbs fn

File: gnark/single/main.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+	Hash               [32]uints.U8
+	Message            emulated.Element[emulated.Secp256k1Fr]
+	Signature          ecdsa.Signature[emulated.Secp256k1Fr]
+}
+
+func (circuit *tronBlockCircuit) Define(api frontend.API) error {
+	blockPublicKeyXLimbs := circuit.PublicKey.X.Limbs
+	blockPublicKeyYLimbs := circuit.PublicKey.Y.Limbs
+
+	// TODO: Try to optimize if possible
```
Comment (alexhooketh) [inline gnark/single/main.go:34]: what's to be optimized here?

**Review by alexhooketh — COMMENTED** [review]
lgtm

**Comment by alexhooketh** [comment]
ok i can't manage to run this so if it's not working just fix it later

File: gnark/single/main.go (hunk 1 above)
Comment (ziemen4) [reply gnark/single/main.go:34]: Idk if there is a more efficient way

**Pull request merged**